Add plumbing to raise Xcode issues when client-side network issues are detected

whsieh opened the pull request:
#### 498d27b7264195c8850d344b542b62f95946c211
<pre>
Add plumbing to raise Xcode issues when client-side network issues are detected
<a href="https://bugs.webkit.org/show_bug.cgi?id=249897">https://bugs.webkit.org/show_bug.cgi?id=249897</a>
rdar://102778314

Reviewed by Tim Horton.

Raise an Xcode issue when certain network activity is observed (`NetworkIssueReporter::shouldReport`
returns `true`).

* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::reportNetworkIssue):

If `NetworkIssueReporter::shouldReport` returned `true`, send an IPC message to the UI process, with
the request URL.

* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::didReceiveResponse):

Consult `NetworkIssueReporter::shouldReport` upon receiving a request response.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::configurationForSessionID):

Unconditionally set `_skipsStackTraceCapture` to `YES`, for WebKit. Regardless of whether or not
we&apos;re actively running in Xcode (and network issue reporting is runtime-enabled), it&apos;s never
necessary to surface stack traces in the WebKit network process to developers.

* Source/WebKit/Platform/SourcesCocoa.txt:
* Source/WebKit/Platform/cocoa/NetworkIssueReporter.h: Added.
* Source/WebKit/Platform/cocoa/NetworkIssueReporter.mm: Added.
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::reportNetworkIssue):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::prepareToLoadWebPage):

Add a new helper method that adds platform data to a given load parameters, and also creates a new
network issue reporter if needed. See below for call sites.

(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::loadFile):
(WebKit::WebPageProxy::loadDataWithNavigationShared):
(WebKit::WebPageProxy::loadSimulatedRequest):
(WebKit::WebPageProxy::loadAlternateHTML):
(WebKit::WebPageProxy::loadWebArchiveData):

Change several call sites of `addPlatformLoadParameters` to use `prepareToLoadWebPage` instead,
which additionally creates a new network issue reporter if needed when triggwering navigation via
API.

(WebKit::WebPageProxy::reportNetworkIssue):

Report the incoming issue if the page&apos;s `NetworkIssueReporter` has been created.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:%

Canonical link: <a href="https://commits.webkit.org/258419@main">https://commits.webkit.org/258419@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98baa2e3699d36cf0c0f4b1e1bec7704aae806aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101870 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11019 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34944 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111205 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171411 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105850 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11981 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1933 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94277 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108961 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107650 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9167 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92430 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/36950 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23858 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78729 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4605 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25344 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4691 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1785 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10770 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44831 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5776 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6443 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/Platform/cocoa/NetworkIssueReporter.h:
```diff
@@ -0,0 +1,56 @@
+
+private:
+    HashSet<String> m_reportedHosts;
+    void* m_stacktrace { nullptr };
```
Given WTF::StackTrace I think we consider this two words and thus the T should be capitalized.

hortont424 replied:
Ditto _skipsStackTraceCapture

whsieh replied:
Fixed!

hortont424 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -1634,6 +1637,8 @@ void WebPageProxy::loadDataWithNavigationShared(Ref<WebProcessProxy>&& process,
     process->send(Messages::WebPage::LoadData(loadParameters), webPageID);
     process->startResponsivenessTimer();
+
+    updateNetworkIssueReporterIfNeeded();
```
At what point do we split some of these things out into a shared function? Seems hairy to expect everyone to hit all the loading paths every time.

hortont424 replied:
Also is "update" really "createNew"?

whsieh replied:
> At what point do we split some of these things out into a shared function? Seems hairy to expect everyone to hit all the loading paths every time.

Good question! Maybe we should have a new helper method like `prepareToLoadWebPage(LoadParameters&)`; we could then move this new logic into that method, along with a call to `addPlatformLoadParameters()`, which we currently call everywhere that we also `createNewNetworkIssueReporterIfNeeded`.

> Also is "update" really "createNew"?

I like that better too — renamed to `createNewNetworkIssueReporterIfNeeded()`.

whsieh committed:
Add plumbing to raise Xcode issues when client-side network issues are detected
https://bugs.webkit.org/show_bug.cgi?id=249897
rdar://102778314

Reviewed by Tim Horton.

Raise an Xcode issue when certain network activity is observed (`NetworkIssueReporter::shouldReport`
returns `true`).

* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::reportNetworkIssue):

If `NetworkIssueReporter::shouldReport` returned `true`, send an IPC message to the UI process, with
the request URL.

* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::didReceiveResponse):

Consult `NetworkIssueReporter::shouldReport` upon receiving a request response.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::configurationForSessionID):

Unconditionally set `_skipsStackTraceCapture` to `YES`, for WebKit. Regardless of whether or not
we're actively running in Xcode (and network issue reporting is runtime-enabled), it's never
necessary to surface stack traces in the WebKit network process to developers.

* Source/WebKit/Platform/SourcesCocoa.txt:
* Source/WebKit/Platform/cocoa/NetworkIssueReporter.h: Added.
* Source/WebKit/Platform/cocoa/NetworkIssueReporter.mm: Added.
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::reportNetworkIssue):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::prepareToLoadWebPage):

Add a new helper method that adds platform data to a given load parameters, and also creates a new
network issue reporter if needed. See below for call sites.

(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::loadFile):
(WebKit::WebPageProxy::loadDataWithNavigationShared):
(WebKit::WebPageProxy::loadSimulatedRequest):
(WebKit::WebPageProxy::loadAlternateHTML):
(WebKit::WebPageProxy::loadWebArchiveData):

Change several call sites of `addPlatformLoadParameters` to use `prepareToLoadWebPage` instead,
which additionally creates a new network issue reporter if needed when triggwering navigation via
API.

(WebKit::WebPageProxy::reportNetworkIssue):

Report the incoming issue if the page's `NetworkIssueReporter` has been created.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:%

Canonical link: https://commits.webkit.org/258419@main
